Fix mixing `Symbol`s and `String`s in `cols`

pdeffebach opened the pull request:
Fixes #181 

This was an easy fix with easy tests to add. 

In the future this might get re-factored to allow for multiple column inputs in `cols`, but not today.

pdeffebach commented on README.md:
```diff
@@ -136,8 +136,11 @@ df = DataFrame(A = 1:3, B = [2, 1, 2])
 df2 = @transform(df, C = :B - cols(nameA))
 
+nameA_string = "A"
+df2 = @transform(df, C = :B - cols(nameA_string))
```
```suggestion
df3 = @transform(df, C = :B - cols(nameA_string))
```

bkamins commented:
what would happen if you passed `1` or `'1'` or `missing` instead of `Symbol` or `AbstractString`?

pdeffebach commented:
Thanks for this. Indeed my solution is too heavy-handed. 

```
df = DataFrame(a = 1)
```

* `1`: error, it looks up the symbol `:1`
* `'1'` same as above
* 'a': works, it looks up the symbol `:a` even though this is not a valid column selector
* `missing`: error, looks up symbol `:missing` (which is very weird)

Is there an internal DataFrames function I should use to sanitize this input?

pdeffebach commented:
This is actually going to be quite hard to support everything, since the implementation current is agnostic about the data frame that is actually being called. I will think on this more.

bkamins commented:
you can check if `source` is `ColumnIndex` or `MultiColumnIndex` maybe (both defined in DataFrames.jl)?

pdeffebach commented:
Thanks I will do that. I was also thinking

```
_names(df)[DataFrames.getindex(df, :a)]
```

or similar. 

The problem is that the current implementation only works with the expression `y = f(:x)` and doesn't rely on the data frame at all, and it's what makes something like 

```
DataFrames.transform(df, @cols y = f(:x))
```

a possibility.

I would have to break this correspondence to fix this. Perhaps it is worth it.

pdeffebach commented:
Here is a very ad-hoc solution

1. Define a function which takes in a dataframe and a vector of column identifiers. It then replaces all of the elements in the vector with integer column positions. 
2. `gensym` the name of the data frame, pass this symbol to the `fun_to_vec` function and write an expression which calls the above function when it returns.
3. Evaluate this expression, which works because we have passed around the relevant `gensym()`-ed symbol. 

Then the actual `transform` call is  

```
_df = df # a gensym()
transform(_df, DataFrames.repair_source(_df, [:a, "a")] => _f => :y)
```

I don't like this solution. It mixes run-time with parse time behavior. It also adds complexity that I don't want to add yet until I have figured out how multi-valued `cols` arguments will work. 

I would rather disallow integers. Disallowing integers means we can call a small function that just converts strings to symbols. If the user wants to use integers they can call `names(df)[4]` to get the 4th column. Thankfully this is easy to do inside `cols`. 

This is a breaking change, as currently you can do `cols(4)` in `@transform`. 


Details:

```
function repair_source(df, x)
    [xi isa Integer ? xi : DataFrames.columnindex(df, xi) for xi in x]
end

function transform_helper(x, args...)
    _df = gensym()
    t = (fun_to_vec(arg, _df) for arg in args)

    for ti in t
        @show ti
    end

    quote
        $_df = $x
        $DataFrames.transform($_df, $(t...))
    end
end
```

bkamins commented:
With `DataFrames.columnindex` just remember that `0` is returned when a column is missing.

pdeffebach commented:
Do you think the solution is viable? Or do you think it is too ad-hoc. 

I think it breaks the "one-to-one" mental model between DataFramesMeta and DataFrames a bit too much by introdutcing the extra feature. 

Do you think this is a valid concern? Do you think users *really* want to do `@transform(df, y = :x + cols(1))`?

bkamins commented:
I would start with a usual question: what is a contract for `cols`?
I assumed it is that `cols(something)` is converted to `df[!, something]` verbatim.
Is this correct?
If yes then we do not need to do anything with what is inside `cols` - just pass it verbatim and let DataFrames.jl either resolve it or error.

> Do you think users `really` want to do `@transform(df, y = :x + cols(1))`?

No, but they might potentially want e.g.:
```
for i in 1:5
    @transform(df, y = :x + cols(i)) # of course something more useful here
end
```

pdeffebach commented:
I guess after the re-design, I would view `@transform` as a way to construct a `source => fun => dest` expression. So the contract with `cols` is "something that goes into the `source`. 

If this is the case then `:x + cols(1)` is clearly invalid, since `[:x, 1]` is an error. On the other hand `:x + cols("y")` should error, but is a small deviation. Importantly, you can reason about the resulting expression *without* knowing much about the data frame you are acting on.

bkamins commented:
Ah - right. I thought about the redesign only after I have commented. In this case, if you want to allow mixing `Symbol` and `AbstractString`, instead of `Symbol.(...)` I would write `(x -> x isa AbstractString ? Symbol(x) : x).(...)`

pdeffebach commented:
I will check if they are all the same type, and if so, return the vector. 

The only mix we allow is `[:x, "y"]`, as a special case. 

This way people can do `y = cols(4) + cols(5)`. This will be easy to reason about. If people complain, then I will re-think the design.

pdeffebach commented:
I have updated the implementation and think this is a reasonable fix. 

This change does *not* apply to `@with`-based macros, though, including `@byrow`. Perhaps I should enforce the same rules for consistency, even though the implementation is fundamentally different.

pdeffebach commented:
I will restrit `@with`, then, as well.

Maybe in the future DataFrames can expose its API for doing this process so `@with` can use it directly or other packages can opt-in.

pdeffebach commented:
Ready for a review! The `@with` code has been changed to not mix integers with other types.

bkamins commented on README.md:
```diff
@@ -136,8 +136,11 @@ df = DataFrame(A = 1:3, B = [2, 1, 2])
 nameA = :A
 df2 = @transform(df, C = :B - cols(nameA))
 
+nameA_string = "A"
```
maybe make it a `julia> ....` and then print output (like we do in docstrings in DataFrames.jl)? It can be a separate PR of course later

pdeffebach replied:
yeah I'm procrastinating some docs changes. I'm going to move things to a real documentation system soon so I won't be worried about the diffs.

bkamins commented on src/DataFramesMeta.jl:
```diff
@@ -139,6 +141,16 @@ end
+    if isempty(x) || isconcretetype(eltype(x))
+        return x
+    elseif all(t -> t isa Union{AbstractString, Symbol}, x)
+        return [xi isa AbstractString ? Symbol(xi) : xi for xi in x]
```
I think this is better for two reasons:
1. `Symbol(:x)` is no-op so no need for `if`
2. if `x` is not a vector then comprehension will convert it to a vector (but maybe you want it? How do you e.g. handle `Tuple`? Or maybe we should check that we got a vector?)

```suggestion
        return Symbol.(x)
```

pdeffebach replied:
Thanks. 

w.r.t vector inputs, I'm using a `Vector` of column references in `@where` even though it's not strictly necessary. This is the right thing to do to avoid too much compilation.

bkamins reviewed (approved):
Looks good except the comments left.

pdeffebach commented:
Requesting a review from @nalimilan before merging.

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -139,6 +141,16 @@ end
 
 fun_to_vec(kw::QuoteNode) = kw
 
+function repair_source(x)
```
Is this appropriate? (It helps understanding the code.)
```suggestion
function repair_source(x::AbstractVector)
```

Could also use a more specific name like `to_symbols` or something similar.

pdeffebach replied:
I added the type annotation. 

However I think the name is good since it doesn't convert everything to `Symbol`s. If given a vector of integers, it leaves it untouched.

nalimilan commented on src/DataFramesMeta.jl:
```diff
@@ -163,15 +175,14 @@ function with_helper(d, body)
     membernames = Dict{Any, Symbol}()
     funname = gensym()
     body = replace_syms!(body, membernames)
-    if isempty(membernames)
```
Isn't this branch relevant anymore?

pdeffebach replied:
I don't think so. That code is pre 0.6. My guess is splatting empty collections wasn't allowed back then or something. All tests pass, even ones with empty arguments. 

Plus, this branch is a source of inconsistent behavior, see #184

nalimilan commented on test/byrow.jl:
```diff
@@ -99,13 +98,6 @@ end
     end
     @test df2 == DataFrame(A = [2, 1, 2], B = [2, 1, 2])
 
-    n = 1
-    df2 = @byrow df begin
-        cols(n) = :B
-    end
-    @test df2 == DataFrame(A = [2, 1, 2], B = [2, 1, 2])
```
Isn't this still allowed?

pdeffebach replied:
No, because of the change with `@with`. We now disallow mixing `:x` and `2` inside the inputs to the anonymous function. 

and `@byrow` is just 

`@with df begin 
    for row in 1:nrow(df)
        :x[row] = :y[row]
    end
end
```

The distinction between LHS and RHS isn't present in `@with` the way it is in `@transform`.

pdeffebach commented:
I have updated the readme with a complete discussion of the behavior and updated NEWS.md. 

I also renamed `repair_source` to `make_source_concrete` for clarity.

pdeffebach commented:
Bumping this, I thin kthe docs are good

bkamins commented:
Let us wait for @nalimilan to comment as he had some questions here.

bkamins commented:
Thank you!
